### PR TITLE
cloud test runner

### PR DESCRIPTION
**Pull request by norlandrhagen**

### CODE REVIEW

**Comment by norlandrhagen** [comment]
https://github.com/carbonplan/cmip6-downscaling/issues/105

File: cmip6_downscaling/data/cmip.py [hunk 1]
```diff
@@ -230,21 +231,27 @@ def get_gcm(
...
-            return_type='xr',
-        ).isel(time=0)
-    else:
-        gcm_grid = gcm_ds.isel(time=0)
+    with dask.config.set(**{'array.slicing.split_large_chunks': False}):
+
+        if gcm_ds is None:
+            """Silences the /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/indexing.py:1228: PerformanceWarning: Slicing is producing a large chunk. To accept the large
+            chunk and silence this warning, set the option >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):"""
+            assert gcm_name is not None, 'one of gcm_ds or gcm_name has to be not empty'
```
Comment (jhamman) [inline cmip6_downscaling/data/cmip.py:248]: turn this into an exception.

```suggestion
            if gcm_name is not None:
                raise ValueError('one of gcm_ds or gcm_name has to be not empty')
```

I think I'd actually catch this before dropping into the if statement. I.e. check if both arguments are supplied.

File: cmip6_downscaling/data/cmip.py [hunk 2]
```diff
@@ -230,21 +231,27 @@ def get_gcm(
...
-            source_ids=gcm_name,
-            return_type='xr',
-        ).isel(time=0)
-    else:
-        gcm_grid = gcm_ds.isel(time=0)
+    with dask.config.set(**{'array.slicing.split_large_chunks': False}):
+
+        if gcm_ds is None:
+            """Silences the /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/indexing.py:1228: PerformanceWarning: Slicing is producing a large chunk. To accept the large
+            chunk and silence this warning, set the option >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):"""
```
Comment (jhamman) [inline cmip6_downscaling/data/cmip.py:247]: multi-line strings are executed at run time and since this is just a comment, the best practice is to just write as a comment.

```suggestion
            # Silences the /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/indexing.py:1228: PerformanceWarning: Slicing is producing a large chunk. To accept the large
            # chunk and silence this warning, set the option >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):"""
```

File: cmip6_downscaling/methods/bcsd.py [hunk 3]
```diff
@@ -168,25 +177,37 @@ def get_spatial_anomalies(
...
     )
-    # calculate the difference between the actual obs (with finer spatial heterogeneity)
-    # and the interpolated coarse obs this will be saved and added to the
-    # spatially-interpolated coarse predictions to add the spatial heterogeneity back in.
+    coarse_obs_interpolated_rechunked_ds = xr.open_zarr(coarse_obs_interpolated_rechunked_path)
+    obs_rechunked_ds = xr.open_zarr(obs_rechunked_path)
+
+    # # calculate the difference between the actual obs (with finer spatial heterogeneity)
+    # # and the interpolated coarse obs this will be saved and added to the
+    # # spatially-interpolated coarse predictions to add the spatial heterogeneity back in.
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:208]:
```suggestion
    # calculate the difference between the actual obs (with finer spatial heterogeneity)
    # and the interpolated coarse obs this will be saved and added to the
    # spatially-interpolated coarse predictions to add the spatial heterogeneity back in.
```

File: cmip6_downscaling/methods/bcsd.py [hunk 4]
```diff
@@ -168,25 +177,37 @@ def get_spatial_anomalies(
     seasonal_cycle_spatial_anomalies : xr.Dataset
         Spatial anomaly for each month (i.e. of shape (nlat, nlon, 12))
     """
-    # Regrid coarse observation dataset to the spatial scale of the raw obs
-    coarse_obs_interpolated, _ = regrid_dataset(
-        ds=coarse_obs,
-        ds_path=None,
-        target_grid_ds=obs_ds.isel(time=0),
-        variable=variable,
+    print(kwargs)
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:180]:
```suggestion
```

File: cmip6_downscaling/methods/bcsd.py [hunk 5]
```diff
@@ -338,31 +360,27 @@ def return_gcm_predict_rechunked(
         predict_period=predict_period,
         bbox=bbox,
     )
-
+    # opts
+    """
+    Q:'s What are the chunks of gcm_train_subset_full_time_ds? Should this just be full_time instead?
+    1. rechunk try: L353
+    2. time part of subset_dataset, after rechunking. Time slicing from subset_dataset func has to happen post-rechunker.
+    """
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:368]: 1. use comment instead of string
2. hopefully this is fully addressed now and can be removed

File: cmip6_downscaling/methods/bcsd.py [hunk 6]
```diff
@@ -338,31 +360,27 @@ def return_gcm_predict_rechunked(
...
-    }
-    gcm_predict_rechunked_ds, _ = rechunk_zarr_array(
-        gcm_predict_ds_subset,
-        zarr_array_location=None,
-        variable=variable,
-        chunk_dims=matching_chunks_dict,
-        max_mem='1GB',
+    del gcm_predict_ds_subset[variable].encoding['chunks']
+    gcm_predict_rechunked_ds = rechunk_zarr_array_with_caching(
+        gcm_predict_ds_subset.chunk({'time': 1500}),
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:378]: add a comment explaining why the `.chunk` is needed here

File: cmip6_downscaling/methods/bcsd.py [hunk 7]
```diff
@@ -431,9 +449,33 @@ def fit_and_predict(
...
+
+def rechunked_spatial_anomalies_full_time(spatial_anomalies_ds: xr.Dataset, **kwargs) -> xr.Dataset:
+
+    rechunked_spatial_anomalies = rechunk_zarr_array_with_caching(
+        spatial_anomalies_ds, chunking_approach='full_time', max_mem='2GB'
+    )
+    print('rechunked_spatial_anomalies:')
+    print(rechunked_spatial_anomalies)
+    print(rechunked_spatial_anomalies.chunks)
+    return rechunked_spatial_anomalies
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:473]: both of these functions only include one function call in them. Are they needed or could they be replaced inline calls to `rechunk_zarr_array_with_caching`

File: cmip6_downscaling/methods/bcsd.py [hunk 8]
```diff
@@ -470,13 +512,45 @@ def postprocess_bcsd(
...
-    bcsd_results_ds = y_predict_fine.groupby("time.month") + spatial_anomalies_ds
-    delete_chunks_encoding(bcsd_results_ds)
-    bcsd_results_ds = bcsd_results_ds.chunk({'time': 30})
-    return bcsd_results_ds
+    del bcsd_results_ds['month'].encoding['chunks']
+
+    print('bcsd_results_ds:')
+    print(bcsd_results_ds)
+    print(bcsd_results_ds.chunks)
+    """[2022-03-05 05:45:09+0000] INFO - prefect.TaskRunner | <xarray.Dataset>
```
Comment (jhamman) [inline cmip6_downscaling/methods/bcsd.py:532]: remove string and print statements

File: cmip6_downscaling/runtimes.py [hunk 9]
```diff
@@ -211,13 +211,16 @@ def run_config(self) -> RunConfig:
...
+        return LocalExecutor()
+
+    # def executor(self) -> Executor:
+    #     return DaskExecutor(
+    #         cluster_kwargs={
+    #             'resources': {'TASKSLOTS': 1},
+    #             'n_workers': self._n_workers,
+    #             'threads_per_worker': self._threads_per_worker,
+    #         }
+    #     )
```
Comment (jhamman) [inline cmip6_downscaling/runtimes.py:223]: revert this change before merging

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 10]
```diff
@@ -62,48 +68,111 @@
...
 )
+
+
 get_coarse_obs_task = task(
     get_coarse_obs,
     tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
-    retry_delay=timedelta(seconds=10),
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=10),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:77]: if these are no longer needed, you can remove them

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 11]
```diff
@@ -62,48 +68,111 @@
...
+    target_grid_obs_ds_task,
+)
+
+
+interpolated_obs_task = task(
+    regrid_ds,
     tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
+    # max_retries=5,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:95]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 12]
```diff
@@ -62,48 +68,111 @@
...
     ),
-    target=make_spatial_anomalies_path,
+    target=make_interpolated_prediction_path_full_space,
+)
+
+rechunked_interpolated_prediciton_task_full_time_task = task(
+    rechunked_interpolated_prediciton_task_full_time,
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=10),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:120]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 13]
```diff
@@ -62,48 +68,111 @@
...
+    get_spatial_anomalies,
+    log_stdout=True,
+    # max_retries=5,
+    # retry_delay=timedelta(seconds=5),
+    result=XpersistResult(
+        intermediate_cache_store,
+        serializer="zarr.zarr",
+        serializer_dump_kwargs=serializer_dump_kwargs,
+    ),
+    target=make_spatial_anomalies_path,  # "flow-outputs/prefect_intermediates/spatial_anomalies_test/ERA5/tasmax/-90.0_90.0_-180.0_180.0/1981_2010/128x256_gridsize_14_14_llcorner_-88_-180.zarr"#make_spatial_anomalies_path"
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:139]: is this needed?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 14]
```diff
@@ -62,48 +68,111 @@
...
+)
+
+
+rechunked_spatial_anomalies_full_time_task = task(
+    rechunked_spatial_anomalies_full_time,
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=10),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:147]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 15]
```diff
@@ -62,48 +68,111 @@
...
+
+
 return_coarse_obs_full_time_task = task(
     return_coarse_obs_full_time,
-    tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
-    retry_delay=timedelta(seconds=5),
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:161]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 16]
```diff
@@ -62,48 +68,111 @@
...
 
+
 return_gcm_train_full_time_task = task(
     return_gcm_train_full_time,
-    tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
-    retry_delay=timedelta(seconds=5),
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:175]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 17]
```diff
@@ -114,9 +183,9 @@
 
 return_gcm_predict_rechunked_task = task(
     return_gcm_predict_rechunked,
-    tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
-    retry_delay=timedelta(seconds=5),
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:188]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 18]
```diff
@@ -128,19 +197,21 @@
 fit_and_predict_task = task(
     fit_and_predict,
     log_stdout=True,
+    # tags=['dask-resource:TASKSLOTS=1'],
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:202]: remove?

File: cmip6_downscaling/workflows/bcsd_flow.py [hunk 19]
```diff
@@ -128,19 +197,21 @@
...
     target=make_bias_corrected_path,
 )
-
 postprocess_bcsd_task = task(
     postprocess_bcsd,
     tags=['dask-resource:TASKSLOTS=1'],
-    max_retries=10,
-    retry_delay=timedelta(seconds=5),
+    # max_retries=10,
+    # retry_delay=timedelta(seconds=5),
```
Comment (jhamman) [inline cmip6_downscaling/workflows/bcsd_flow.py:214]: remove?

**Comment by norlandrhagen** [comment]
Status:
Working at least through `postprocess`

Notes: Above comments from jhamman  need to be address. XESMF error hasn't appeared in any of the recent test runs.
